### PR TITLE
Add encryption for passwords

### PR DESCRIPTION
**Pull request by Quentin-L-Inge**
This pull request introduces a new feature for securely creating and storing password entries, along with significant improvements to encryption and database handling. The main changes include implementing a password creation workflow, enhancing encryption with authentication tags, and updating the database schema to support these improvements. Additionally, unit tests were added to ensure the reliability of the encryption and decryption processes.

### New Feature: Password Entry Creation
* Added a `CreateEntry` component (`components/CreateEntry.vue`) that allows users to create password entries with a UI for input and a button to trigger the creation process. The password is sent to the server for encryption and storage.
* Integrated the `CreateEntry` component into the main page (`pages/index.vue`).

### Encryption Enhancements
* Updated the `encrypt` and `decrypt` functions in `encrypt-password.ts` and `decrypt-password.ts` to include an authentication tag (`authTag`) for AES-256-GCM encryption, improving data integrity and security. [[1]](diffhunk://#diff-3cbde476f121db1c5740e08ac33afb1efb5fc869680e84cc5828df190897632bL8-R16) [[2]](diffhunk://#diff-c1dc22d3ec5a44e613bb8131ce8848398422f1af3b9f9c6f89547a45b5990c6dL3-R7)
* Enhanced the `encryptEncryptionKey` and `decryptEncryptionKey` functions to handle the `authTag` for encrypting and decrypting the encryption key.

### Database Updates
* Modified the database schema to include an `auth_tag` column in the `password` table for storing the authentication tag. [[1]](diffhunk://#diff-cb424de399435c99f87383854f67321b45b8ec5ceed2e5f79445a5f4fc53646bR1) [[2]](diffhunk://#diff-3a32612d4838b26e1af163134cf2231e6b5e880ddc79f537a9a4aa42ccda1804R54)
* Added a migration script (`0001_complex_marrow.sql`) to apply the schema changes.

### Backend Improvements
* Implemented a new API endpoint (`server/api/passwords/create-entry.post.ts`) for handling password entry creation, including validation, encryption, and database insertion.
* Updated the `generate-master-key` endpoint to unseal the database after generating encryption keys, ensuring proper initialization.

### Unit Tests
* Added comprehensive unit tests for the encryption and decryption functions, covering various scenarios such as empty passwords, special characters, and long passwords.
* Updated existing tests for `encryptEncryptionKey` to validate the inclusion of the `authTag`. [[1]](diffhunk://#diff-bf7e9dfb5f720192b5d78314c0a77bae662c3d37fe310710f8a9507860a5ec65L9-R28) [[2]](diffhunk://#diff-bf7e9dfb5f720192b5d78314c0a77bae662c3d37fe310710f8a9507860a5ec65R39-R43)

### CODE REVIEW

File: server/utils/encryption/encryption-key.ts [hunk 1]
```diff
@@ -1,23 +1,29 @@
 import { Buffer } from 'node:buffer'
 import crypto from 'node:crypto'
 
-export function encryptEncryptionKey(encryptionKey: Uint8Array, masterKey: Uint8Array): { iv: string, encrypted: string } {
-  const iv = crypto.getRandomValues(new Uint8Array(16))
+export function encryptEncryptionKey(encryptionKey: Uint8Array, masterKey: Uint8Array): { iv: string, encrypted: string, authTag: string } {
```
Comment (vianneybacoup) [inline server/utils/encryption/encryption-key.ts]: ça pourrait bien d'avoir un classe "EncryptedPassword" qui contient tes trois valeurs (iv, encrypted et authTag)
Comme ça tu pourras lui ajouter des méthodes qui colle avec sa logique, comme "decryptEncryptionKey" du dessous

File: server/api/passwords/create-entry.post.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  userPassword: z.string().min(1, 'Password is required'),
+})
+
+export default defineEventHandler(async (event) => {
+  // Vérifier que l'utilisateur est authentifié
+  const session = await auth.api.getSession(toWebRequest(event))
+  if (!session || !session.user) {
+    setResponseStatus(event, 401)
+    return {
+      success: false,
```
Comment (vianneybacoup) [inline server/api/passwords/create-entry.post.ts]: 401 == error, pas sûr que le "success: false" soit nécessaire
Pareil error "Unauthorized" ça ajoute pas grand chose au 401

Par contre "User not authenticated" ça ajoute autre chose

File: server/utils/encryption/encryption-key.ts [hunk 3]
```diff
@@ -1,23 +1,29 @@
...
-    iv: ivHex,
-    encrypted: encryptedHex,
+    iv: iv.toString('hex'),
+    encrypted: encrypted.toString('hex'),
+    authTag: authTag.toString('hex'),
   }
 }
 
-export function decryptEncryptionKey(encryptedKey: string, masterKey: Uint8Array, iv: string): Uint8Array {
+export function decryptEncryptionKey(encryptedKey: string, masterKey: Uint8Array, iv: string, authTag: string): Uint8Array {
```
Comment (vianneybacoup) [inline server/utils/encryption/encryption-key.ts]: Beaucoup de "Uint8Array" ça serait propre d'avoir un type à part aussi "EcnryptionKey" ou qqchose plus adéquat

File: tests/unit/encrypt-password.test.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
+import { describe, expect, it } from 'vitest'
+import { decrypt } from '~/server/utils/encryption/decrypt-password'
+import { encrypt } from '~/server/utils/encryption/encrypt-password'
+
+describe('encrypt et decrypt password', () => {
+  // Test de la structure retournée
+  it('encrypt devrait retourner un objet avec les propriétés iv, encryptedPassword et authTag', () => {
+    const encryptionKey = new Uint8Array(32).fill(1) // Clé de test
+    const password = 'MonMotDePasse123!'
```
Comment (vianneybacoup) [inline tests/unit/encrypt-password.test.ts]: pas très sécurisé comme mot de passe :D

File: tests/unit/encrypt-password.test.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    const result = encrypt(password, encryptionKey)
+
+    expect(result).toHaveProperty('iv')
+    expect(result).toHaveProperty('encryptedPassword')
+    expect(result).toHaveProperty('authTag')
+  })
+
+  // Test du format
+  it('iv, authTag devraient être des tableaux de nombres et encryptedPassword une chaîne hexadécimale', () => {
```
Comment (vianneybacoup) [inline tests/unit/encrypt-password.test.ts]: Hmm je pense que c'est un détail d'implémentation.
En fait que ce soit d'autres types ou pas, du moment que t'arrives à les utiliser pour chiffrer/déchiffrer, et ben on s'en fou un peu de comment ils sont

File: tests/unit/encrypt-password.test.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    expect(result1.iv).not.toEqual(result2.iv)
+  })
+
+  // Edge case: mot de passe vide
+  it('devrait fonctionner avec un mot de passe vide', () => {
+    const encryptionKey = new Uint8Array(32).fill(1)
+    const emptyPassword = ''
+
+    const { iv, encryptedPassword, authTag } = encrypt(emptyPassword, encryptionKey)
+    const ivUint8Array = new Uint8Array(iv)
```
Comment (vianneybacoup) [inline tests/unit/encrypt-password.test.ts]: à mettre au début du test, normalement on suit AAA
Arrange (setup)
Act (exec la function)
Assert (check du résultat)

Comment (Quentin-L-Inge) [reply tests/unit/encrypt-password.test.ts]: Je n'ai aucune connaissance en cela, d'où le fait que je n'avais pas fais de test unitaire. Je suis intéressé par une explication.

File: tests/unit/encrypt-password.test.ts (hunk 4 above)
Comment (Quentin-L-Inge) [reply tests/unit/encrypt-password.test.ts]: Comme la plupart des mots de passes :) mais je change ça :)

File: server/utils/encryption/encryption-key.ts (hunk 1 above)
Comment (Quentin-L-Inge) [reply server/utils/encryption/encryption-key.ts]: Je ne suis pas sûr de voir ce que tu veux dire.

Comment (vianneybacoup) [reply server/utils/encryption/encryption-key.ts]: En gros retourner plusieurs valeurs ou un dictionnaire c'est pas ouf
Soit tu passes par un type, soit une classe pour regrouper tout ça

Parce que iv, encrypted et authTag tout seul c'est pas super explicite

Et surtout, ça te permet d'attacher à ces valeurs des méthodes qui leur sont propres
